internal/expr/query.go

```diff
@@ -58,38 +77,80 @@ func (pe *PreparedExpr) Query(args ...any) (ce *QueryExpr, err error) {
-		if !inQuery[t] {
-			// Check if we have a type with the same name from a different package.
-			for _, typeMember := range pe.inputs {
-				if t.Name() == typeMember.outerType().Name() {
-					return nil, fmt.Errorf("parameter with type %q missing, have type with same name: %q", typeMember.outerType().String(), t.String())
+	}
+
+	// Query parameteres.
+	qargs := make([]any, 0)
+	outputs := make([]typeMember, 0)
+	inQuery := make(map[reflect.Type]bool)
+	inCount := 0
+	outCount := 0
+	var sqlStr bytes.Buffer
+	for _, pp := range pe.preparedParts {
+		switch pp := pp.(type) {
+		case *preparedInput:
+			typeMember := pp.inputValue
+			outerType := typeMember.outerType()
+			v, ok := typeValue[outerType]
+			if !ok {
+				argTypes := make([]reflect.Type, 0)
+				for argType, _ := range typeValue {
+					argTypes = append(argTypes, argType)
 				}
+				if err := checkShadowedType(outerType, argTypes); err != nil {
+					return nil, err
+				}
+				return nil, typeMissingError(outerType.Name(), typeNames)
 			}
-			return nil, fmt.Errorf("%s not referenced in query", t.Name())
+			inQuery[outerType] = true
+			var val reflect.Value
+			switch tm := typeMember.(type) {
+			case *structField:
+				val = v.Field(tm.index)
+			case *mapKey:
+				val = v.MapIndex(reflect.ValueOf(tm.name))
+				if val.Kind() == reflect.Invalid {
+					return nil, fmt.Errorf(`map %q does not contain key %q`, outerType.Name(), tm.name)
+				}
+			}
+			qargs = append(qargs, sql.Named("sqlair_"+strconv.Itoa(inCount), val.Interface()))
+
+			sqlStr.WriteString("@sqlair_" + strconv.Itoa(inCount))
+			inCount++
+		case *preparedOutput:
+			for i, c := range pp.queryColumns {
+				sqlStr.WriteString(c.String())
+				sqlStr.WriteString(" AS ")
+				sqlStr.WriteString(markerName(outCount))
+				if i != len(pp.queryColumns)-1 {
+					sqlStr.WriteString(", ")
+				}
+				outCount++
+			}
+			outputs = append(outputs, pp.outputValues...)
+		case *preparedBypass:
+			sqlStr.WriteString(pp.chunk)
 		}
 	}
 
-	// Query parameteres.
-	qargs := []any{}
-	for i, typeMember := range pe.inputs {
-		outerType := typeMember.outerType()
-		v, ok := typeValue[outerType]
-		if !ok {
-			return nil, typeMissingError(outerType.Name(), typeNames)
+	for argType, _ := range typeValue {
+		if !inQuery[argType] {
+			return nil, fmt.Errorf("%s not referenced in query", argType.Name())
 		}
-		var val reflect.Value
-		switch tm := typeMember.(type) {
-		case *structField:
-			val = v.Field(tm.index)
-		case *mapKey:
-			val = v.MapIndex(reflect.ValueOf(tm.name))
-			if val.Kind() == reflect.Invalid {
-				return nil, fmt.Errorf(`map %q does not contain key %q`, outerType.Name(), tm.name)
-			}
+	}
+
+	return &QueryExpr{outputs: outputs, sql: sqlStr.String(), args: qargs}, nil
+}
+
+// checkShadowedType returns an error if a query type and some argument type
+// have the same name but are from a different package.
+func checkShadowedType(queryType reflect.Type, argTypes []reflect.Type) error {
```

Comment:
I like that you have created a function for this. I was little skeptical reading the code because it is only used in one place but I think it aids readability. This way, we have less indentation levels in the above function and, from this function name, it is clear what it does.